src/lazy/encoder/binary/v1_1/flex_uint.rs

```diff
@@ -0,0 +1,229 @@
+use crate::IonResult;
+use std::io::Write;
+
+const BITS_PER_U64: usize = 64;
+const BITS_PER_ENCODED_BYTE: usize = 7;
+
+// Compile-time mapping from number of leading zeros to the number of bytes needed to encode
+const fn init_bytes_needed_cache() -> [u8; 65] {
+    let mut cache = [0u8; 65];
+    let mut leading_zeros = 0usize;
+    while leading_zeros < BITS_PER_U64 {
+        let magnitude_bits_needed = 64 - leading_zeros;
+        cache[leading_zeros] =
+            ((magnitude_bits_needed + BITS_PER_ENCODED_BYTE - 1) / BITS_PER_ENCODED_BYTE) as u8;
+        leading_zeros += 1;
+    }
+    // Special case; 64 leading zeros means the value is zero. We need a byte to represent it.
+    cache[64] = 1;
+    cache
+}
+
+static BYTES_NEEDED_CACHE: [u8; 65] = init_bytes_needed_cache();
+
+/// An Ion 1.1 encoding primitive that represents a variable-length unsigned integer.
+#[derive(Debug)]
+pub struct FlexUInt {
+    value: u64,
+    size_in_bytes: usize,
+}
+
+impl FlexUInt {
+    pub fn new(size_in_bytes: usize, value: u64) -> Self {
+        Self {
+            value,
+            size_in_bytes,
+        }
+    }
+
+    #[inline]
+    pub fn write_u64<W: Write>(output: &mut W, value: u64) -> IonResult<usize> {
```

Comment:
Is there any case where we would write a `FlexUInt` from something smaller than a `u64`? If so, it may be worth adding a second function for `u32` since anything `u32` or smaller can be written without branching.

Reply:
Yes, though I'd like to punt on this for the sake of this PR. Tracking it in #693.